docs(network): specify data structure for transactions (#250)
witnet/witnet-rust: docs/protocol/network/data-structures/transaction.md

# Transaction

In the Witnet network protocol, a `transaction` is formatted as follows:

| Field     |    Type    | Description                                                               |
| --------- | :--------: | ------------------------------------------------------------------------- |
| `version` |   `u32`    | The transaction data format version number                                |
| `inputs`  | `[input]`  | A list of 1 or more transaction inputs                                    |
| `outputs` | `[output]` | A list of transaction outputs                                             |

> | `claims`  | `[claim]`  | A list of 1 or more claims (i.e. spending conditions), one for each input |

Reviewer: The requirement for _1 or more_ inputs and claims is in conflict with the existence of coinbase transactions.

Author: Then, it should be `1 or more transaction outputs` to enforce at least the coinbase. :)

Reviewer: Only implication then is that you can't produce transactions that donate all their value to the miner without changing or burning at least one _satowit_. I'm totally ok with that.

Author: Is it OK as follows?

```table
| Field     |    Type    | Description                                 |
| --------- | :--------: | ------------------------------------------- |
| `version` |   `u32`    | The transaction data format version number  |
| `inputs`  | `[input]`  | A list of transaction inputs                |
| `outputs` | `[output]` | A list of 1 or more transaction outputs     |
| `claims`  | `[claim]`  | A list of claims (i.e. spending conditions) |
```

Inputs and claims may be empty due to the `coinbase` transaction. The claim for mining the block is included in the block itself as leadership proof.

The inputs and claims lists cannot be empty as trasactions require at least to consume an unspent output.

Signature structures are defined in the [Signature] section.

> ## Inputs (aka outpoints)

Reviewer: It's true that all inputs are of the type `outpoint` but I wouldn't dare to assimilate them (just like I wouldn't say `Bakers (aka humans)`)

Generally speaking, `outpoint` is just an alias for a tuple of a transaction's id and the index of one of its outputs. Giving too much protagonism to the `outpoint` concept may render the documentation more confusing and less approachable for people not familiar to it.

> The transaction inputs are also known as outpoints because they refer to an existing unspent transaction output.

Reviewer: Transaction inputs don't necessarily point to unspent transaction outputs. That's only truth for inputs in transactions that are not yet anchored to the chain and that don't point to outputs in other transactions that are equally not yet anchored to the chain.

This paragraph could read:
```suggestion
Transaction inputs are references to outputs from past transactions. They "pull" all the value from those outputs and make it available for being spent by the outputs in the same transaction they are in. This data structure—which pairs a transaction's identifier with the index of one of its outputs—is also known as `outpoint` as it unambiguously points to a unique output from a specific transaction.
```

| `backup_witnesses` |  `u8`  | Number of backup data request witnesses                                                                                         |
| `commit_fee`       | `u64`  | Miner fee for each valid commit output included in the block during the **commit phase**                                        |
| `reveal_fee`       | `u64`  | Miner fee for each valid reveal output included in the block during the **reveal phase**                                        |
| `tally_fee`        | `u64`  | Miner fee for each valid P2PKH output included in the block during the  **tally phase**                                         |
| `time_lock`        | `u64`  | The UTC Unix timestamp after which data request may be executed                                                                 |

The minimum data request reward to be distributed among nodes that agreeed with the consensus is defined as follows:

```math
dr_reward_min = value - (witnesses * commit_fee) - (witnesses * reveal_fee) - (witnesses * tally_fee)
```

During the reveal and tally phase, some committers may not participate in the data request.

> However, the block miner is incentivized to include as many outputs as committers as possible in order to increase the block fees.

Reviewer: ```suggestion
Miners are actually not obliged to include all the reveal transactions and eventually end up assigning rewards to the committers. This is because there is no way for the network to enforce punishment on them for neglecting or trying to conceal those transactions because there's no guarantee that they will be known to them in discrete time or even known whatsoever. 

However, for every of those transactions that they include in a block, they are eligible for collecting special fees as explicitly specified and set aside for them in the original data request output. It is therefore to be expected that miners will include as many of those transactions as known to them as for maximizing their profit.
```

### Commit outputs

| Field        |  Type  | Description                                                                                                |
| ------------ | :----: | ---------------------------------------------------------------------------------------------------------- |

> | `commitment` | `[u8]` | Signed hash of the data request result by an individual node                                               |

Reviewer: Shall we specify that the commitment is nonced to prevent lazy commitment / replaying?

Reviewer 2: Makes sense

| `value`      | `u64`  | Remaining transaction value that will be used as reward to be distributed after consensus has been reached |

The `value` of the commit output depends on the number of committers:

```math
commit_value = (data_request_value / witnesses) - commit_fee
```

### Reveal outputs

| Field   | Type  | Description                                                                                                |
| ------- | :---: | ---------------------------------------------------------------------------------------------------------- |
| `value` | `u64` | Remaining transaction value that will be used as reward to be distributed after consensus has been reached |

The `value` of the reveal output depends on the number of committers that revealed their data request results:

```math
reveal_value = commit_value - reveal_fee
```

## Claims

> As aforementioned, there should be a claim for each input, which provides an spending condition in order to consume an unspent output (UTXO).

Reviewer: ```suggestion
As aforementioned, transactions should include as many claims as inputs. In every transaction, claims provide the material required for satisfying the spending conditions that encumbered the past transaction outputs that the inputs in the transaction are trying to spend (e.g. signatures). Claims and inputs are matched positionally, that is, the first claim is checked against the first input and so forth.
```

> In Witnet, different output types have specific claims:

Reviewer: ```suggestion
In Witnet, different output types implicitly have different spending conditions and therefore the spending transaction must provide specific items in their claims:
```

> - P2PKH claim: demostrate ownership of private key that signs the hash of the transaction.

Reviewer: ```suggestion
- Value transfer claim: prove ownership of a certain private key.
```

> - Data request claim: provide Proof of Eligibility (PoE) and include Public Key as sign of commitment.

Reviewer: ```suggestion
- Data request claim: prove eligibility for resolving the data request and publishing a commitment of the resulting value.
```

- Commit claim: provide a signed reveal message matching a previous `commitment` (hash) and that matches with the previous advertised public key.
- Reveal claim: provide a data request consensus result which is valid according to the consensus script and previous reveals outputs.

> ### P2PKH claim

Reviewer: ```suggestion
### Value transfer claim
Value transfer claims prove ownership of a certain private key by providing a signature of the identifier of the transaction produced with such key and the serialization of the matching public key.
```

| Field        |  Type  | Description                                     |
| ------------ | :----: | ----------------------------------------------- |
| `signature`  | `[u8]` | Signed hash of the transaction                  |
| `public_key` | `[u8]` | Public Key of the P2PKH outpoint to be consumed |

### Data request claim (**commit phase**)

A **data request** output has to be consumed/used by a number of **witnesses**, i.e. there will be as many claims as **witnesses** have been defined.

> 

Reviewer: ```suggestion
For a committer to be able to pledge a share of the reward from the data request, they must provide a _Proof of Eligibility_ (PoE) that proves their eligibility as witnesses for such data request in the current epoch. In addition, for everyone in the network to be able to verify such proof, they must include the public key that matches the private key that produced the PoE.
```

| Field        |  Type  | Description                           |
| ------------ | :----: | ------------------------------------- |
| `poe`        | `[u8]` | Proof of eligibility                  |
| `public_key` | `[u8]` | Public Key used for computing the PoE |

### Commit claim (**reveal phase**)

| Field    |  Type  | Description                                                                                                |
| -------- | :----: | ---------------------------------------------------------------------------------------------------------- |

> | `reveal` | `[u8]` | Signed data request result using the previously advertised public key during the previous **commit phase** |

Reviewer: This is where the nonce used in the commitment should be published.

Author: I think so... i propose to add a `nonce` field of type `u32`. WDYT?
